File: leetcode/368. Largest Divisible Subset.rs
impl Solution {
    pub fn largest_divisible_subset(mut nums: Vec<i32>) -> Vec<i32> {
        nums.sort();
        let mut indexes: Vec<usize> = nums.iter().enumerate().map(|(i, _)| { i }).collect();
        let mut counts = vec![1; nums.len()];
        
        for i in (0..nums.len()) {
            for j in (0..i) {
                if counts[j] + 1 <= counts[i] {
                    continue;
                }
                if nums[i] % nums[j] == 0 {
                    indexes[i] = j;
                    counts[i] = counts[j] + 1;
                }
            }
        }
        
        let mut res = Vec::new();
        let mut idx = 0;
        let mut max = 0;

        for (i, cnt) in counts.into_iter().enumerate() {
            if cnt > max {
                idx = i;
                max = cnt;
            }
        }

        while let Some(&i) = indexes.get(idx) {
            res.push(nums[idx]);
            if idx == i {
                break;
            }
            idx = i;
        }
        
        res
    }
}

File: leetcode/368. Largest Divisible Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums.sort()
        dp = [1] * len(nums)
        last = [i for i in range(len(nums))]

        for i in range(len(nums)):
            for j in range(i):
                if nums[i] % nums[j] == 0:
                    if dp[j]+1 > dp[i]:
                        dp[i] = dp[j]+1
                        last[i] = j
        
        max_len = 0
        max_idx = -1
        for i, l in enumerate(dp):
            if l > max_len:
                max_len = l
                max_idx = i
        res = []
        i = max_idx
        while last[i] != i:
            res.append(nums[i])
            i = last[i]
        res.append(nums[i])
        return res
